cmd/revad/runtime/runner.go

```diff
@@ -0,0 +1,213 @@
+package runtime
+
+import (
+	"context"
+	"errors"
+	"net"
+	"net/http"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/owncloud/reva/v2/pkg/registry"
+	"github.com/owncloud/reva/v2/pkg/rgrpc"
+	"github.com/owncloud/reva/v2/pkg/rhttp"
+	"github.com/rs/zerolog"
+	"golang.org/x/sync/errgroup"
+)
+
+type servers struct {
+	rhttpServer             *rhttp.Server
+	rgrpcServer             *rgrpc.Server
+	gracefulShutdownTimeout int
+	pidFilel                string
+	log                     *zerolog.Logger
+}
+
+type Runner interface {
+	Start() error
+	Stop(ctx context.Context) error
+}
+
+// RunServerWithOptions runs a revad server w/o wacher with the given config file, pid file and options.
+func RunServerWithOptions(mainConf map[string]interface{}, pidFile string, opts ...Option) Runner {
+	options := newOptions(opts...)
+	parseSharedConfOrDie(mainConf["shared"])
+	coreConf := parseCoreConfOrDie(mainConf["core"])
+
+	if err := registry.Init(options.Registry); err != nil {
+		panic(err)
+	}
+
+	host, _ := os.Hostname()
+	log := options.Logger
+	log.Info().Msgf("host info: %s", host)
+
+	// Only initialise tracing if we didn't get a tracer provider.
+	if options.TraceProvider == nil {
+		log.Debug().Msg("no pre-existing tracer given, initializing tracing")
+		options.TraceProvider = initTracing(coreConf)
+	}
+	initCPUCount(coreConf, log)
+
+	server := &servers{
+		rhttpServer: initHTTPServer(mainConf, &options),
+		rgrpcServer: initGRPCServer(mainConf, &options),
+		log:         log,
+		pidFilel:    pidFile,
+	}
+	server.gracefulShutdownTimeout = 30
+	if coreConf.GracefulShutdownTimeout > 0 {
+		server.gracefulShutdownTimeout = coreConf.GracefulShutdownTimeout
+	}
+
+	if server.rhttpServer == nil && server.rgrpcServer == nil {
+		log.Fatal().Msg("nothing to do, no grpc/http enabled_services declared in config")
+	}
+	return server
+}
+
+func (s *servers) Start() error {
+	eg := new(errgroup.Group)
+	if s.rhttpServer != nil {
+		eg.Go(s.startHTTPServer)
+	}
+	if s.rgrpcServer != nil {
+		eg.Go(s.startGRPCServer)
+	}
+
+	err := eg.Wait()
+	if err != nil {
+		s.log.Error().Err(err).Msg("error starting servers")
+		return err
+	}
+	s.log.Info().Msg("all servers started")
+	return nil
+
+}
+
+func (s *servers) Stop(ctx context.Context) error {
```

Comment:
This will need comments.
As it is at the moment, the function will block the thread until either the servers have stopped or the timeout has been reached. This is fine.
Another alternative could be that the function just tells the server to stop, returning immediately.

Just note that the main thread might finish faster than the one calling the `Stop` function, so it's possible there is code that won't be executed. For the most part, it might not be a big deal because they're likely "shutdown" logs that aren't very important, but we might need to be careful.